### PR TITLE
fix(api): validate username with regex

### PR DESCRIPTION
**Pull request by richardlt**
1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

File: sdk/user.go [hunk 1]
```diff
@@ -49,10 +49,17 @@ type AuthentifiedUser struct {
 	Organization string       `json:"organization,omitempty" yaml:"organization,omitempty" cli:"organization" db:"-"`
 }
 
+func IsValidUsername(username string) error {
+	if username == "" || username == "me" || usernameRegex.MatchString(username) {
+		return NewErrorFrom(ErrInvalidUsername, "invalid given username")
```
Comment (sguiheux) [inline sdk/user.go:54]:
```suggestion
		return NewErrorFrom(ErrInvalidUsername, "invalid given username: %s", username)
```

File: engine/api/user/dao.go [hunk 2]
```diff
@@ -129,7 +129,7 @@ func CountAdmin(db gorp.SqlExecutor) (int64, error) {
 
 // Insert a user in database.
 func Insert(ctx context.Context, db gorpmapper.SqlExecutorWithTx, au *sdk.AuthentifiedUser) error {
-	if !sdk.UsernameRegex.MatchString(au.Username) {
+	if err := sdk.IsValidUsername(au.Username); err != nil {
 		return sdk.WithStack(sdk.ErrInvalidUsername)
```
Comment (sguiheux) [inline engine/api/user/dao.go:133]:
```suggestion
		return err
```

File: engine/api/user/dao.go [hunk 3]
```diff
@@ -146,6 +146,10 @@ func Insert(ctx context.Context, db gorpmapper.SqlExecutorWithTx, au *sdk.Authen
 
 // Update a user in database.
 func Update(ctx context.Context, db gorpmapper.SqlExecutorWithTx, au *sdk.AuthentifiedUser) error {
+	if err := sdk.IsValidUsername(au.Username); err != nil {
+		return sdk.WithStack(sdk.ErrInvalidUsername)
```
Comment (sguiheux) [inline engine/api/user/dao.go:150]:
```suggestion
		return err
```

File: engine/api/authentication/local/dao_registration.go [hunk 4]
```diff
@@ -43,7 +43,7 @@ func LoadRegistrationByID(ctx context.Context, db gorp.SqlExecutor, id string) (
 
 // InsertRegistration in database.
 func InsertRegistration(ctx context.Context, db gorpmapper.SqlExecutorWithTx, ur *sdk.UserRegistration) error {
-	if !sdk.UsernameRegex.MatchString(ur.Username) {
+	if err := sdk.IsValidUsername(ur.Username); err != nil {
 		return sdk.WithStack(sdk.ErrInvalidUsername)
```
Comment (sguiheux) [inline engine/api/authentication/local/dao_registration.go:47]:
```suggestion
		return err
```

**Review by sguiheux — CHANGES_REQUESTED** [review]

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#16836.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Review by yesnault — APPROVED** [review]

**Review by fsamin — APPROVED** [review]

**Review by sguiheux — APPROVED** [review]

**Pull request merged**